pkg/git/client.go

```diff
@@ -0,0 +1,197 @@
+package git
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"regexp"
+	"sort"
+	"strings"
+
+	"github.com/cli/safeexec"
+)
+
+var remoteRE = regexp.MustCompile(`(.+)\s+(.+)\s+\((push|fetch)\)`)
+
+type Client struct {
+	command  func(ctx context.Context, name string, args ...string) *exec.Cmd
+	dir      string
+	exeErr   error
+	exePath  string
+	stderr   io.Writer
+	stdin    io.Reader
+	stdout   io.Writer
+	workTree string
+}
+
+type ClientOptions struct {
+	Command  func(ctx context.Context, name string, args ...string) *exec.Cmd
+	Dir      string    // will be passed to commands as --git-dir
+	ExePath  string    // will be resolved if none passed
+	Stderr   io.Writer // will use stderr if none specified
+	Stdin    io.Reader // will use stdin if none specified
+	Stdout   io.Writer // will use stdout if none specified
+	WorkTree string    // will be passed to commands as --work-tree
+}
+
+func NewClient(opts *ClientOptions) Client {
+	if opts == nil {
+		opts = &ClientOptions{}
+	}
+	if opts.Stderr == nil {
+		opts.Stderr = os.Stderr
+	}
+	if opts.Stdin == nil {
+		opts.Stdin = os.Stdin
+	}
+	if opts.Stdout == nil {
+		opts.Stdout = os.Stdout
+	}
+	return Client{
+		command:  opts.Command,
+		dir:      opts.Dir,
+		exePath:  opts.ExePath,
+		stderr:   opts.Stderr,
+		stdin:    opts.Stdin,
+		stdout:   opts.Stdout,
+		workTree: opts.WorkTree,
+	}
+}
+
+func (c *Client) Command(ctx context.Context, args ...string) (*exec.Cmd, error) {
```

Comment:
My advice would be to not make this public, otherwise the end caller has to be concerned with API details of git on the command-line, including the names of flags, etc. Ideally, we would have completely hidden git's own API behind our own Go API that just exposes the pieces of functionality that we need for our own purposes.

Reply:
Without having this exported then the only method the client has is `Remotes`. I feel like this method allows for ultimate flexibility when interacting with `git` and then we don't need to anticipate or implement every one of our users needs, we can just tell them to utilize `Command` and everything will be taken care of them. Going down the route of implementing every `git` command with every option they allow seems like a lot of work, and there are other `git` packages out there that do that which extension authors can choose to use instead if they like that interface.